design/milestone_1_2_design_doc.md

```diff
@@ -0,0 +1,430 @@
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ----------------- |
+| 1    | Keep using only **namespace** and/or **service account** granularities. | \- Since Secrets Provider runs as a separate entity, we suggest creating a dedicated service account for Secrets Provider.<br />That way, the **service account** granularity serves as an exact match and **Job** granularity is redundant<br />- No code changes | \- Service account might be reused, which allows other apps to authenticate using Secrets Provider host<br />- Break application identity convention | Free              |
+| 2    | Add support for **Job** granularity in Conjur                | Allows specific identification of the Secrets Provider **Job**, which enhances security | \- Costly; requires code changes, verify compatibility, tests, docs<br />- Redundant if **service account** serves only the Secrets Provider | 10 days           |
+
+*Decision*: Solution #1, use **namespace** and **service account** to define the Secrets Provider host in Conjur policy. We would recommend to the customer to create a dedicated service account for Secrets Provider.
+
+#### Customer experience
+
+Deployment for the Secrets Provider will be done using Helm. That way, if the 
+
+1. Configure K8s Secrets with ***conjur-map\*** metadata defined
+
+2. Create `values.yaml` that will contain the following parameters and their defaults:
+
+   2.1. Service Account for Secrets Provider (default: `secrets-provider-account`)
+
+   2.2. Role (default: `secrets-provider-role`)
+
+   2.3. RoleBinding Name (default: `secrets-provider-role-binding`)
+
+   2.4. `K8S_SECRETS` (default: `[]`)
+
+   *When parameters are not configured by customer, their defaults are used*
+
+3. Install Helm Chart for the Secrets Provider
+
+We will supply the customer with a  `values.yaml` file where all default parameters are collected. Our `values.yaml` will resemble the following:
+
+```
+# K8s permissioning defaults
+permission:
+	serviceAccountName: secrets-provider-account
+	roleName: secrets-provider-role
+	roleBinding: secrets-provider-role-binding
+	rules:
+		resources: secrets
+		verbs: [ "get", "update" ]
+
+# Secrets Provider pod defaults
+secrets-provider:
+	image: cyberark/secrets-provider-for-k8s
+	name: cyberark-secrets-provider
+	namespace: default-namespace
+	ttlSecondsAfterFinished: 0	# How long the Job should stay alive even after finishing task. Used for logging.
+	
+# Secrets Provider environment variable defaults
+secrets-provider-environment:
+	conjurApplianceUrl: https://conjur-follower.default.svc.cluster.local
+	conjurAuthnUrl: https://conjur-follower.default.svc.cluster.local/authn-k8s/authn-default
+	conjurAuthnLogin: host/conjur/authn-k8s/authn-default/default
+	conjurAccount: default
+	conjurSslCertificate: conjur-master-ca-env
```

Comment:
We should consider to retrieve certificate as part if start script that issues secret provider.
This may improve customer experience

Reply:
@InbalZilberman what do you think? If we have a small script that runs before the secrets provider that can drop down into the follower container and fetch the ssl certificate and fill the configMap? That way the customer won't need to do it manually. 

Note this adds time to delivery

Reply:
I suggest not event to fill the configmap, but to pass it as an environment var directly to secret provider.
It will be exactly the same like we provided PVWA certificate to PAS Replicator in DAPNG101

Reply:
I have a related question - I didn't see in the `ConfigMap` creation in the helm chart. Do we expect the customer to create it?

How about letting the customer add the ssl_cert in `values.yaml` and adding this to the helm:
```
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: conjur-master-ca-env
  labels:
    app: test-env
data:
  ssl-certificate: |
    {{ .Values.secrets-provider-environment.conjurSslCertificate }}
```

this way the ssl_cert name (`conjur-master-ca-env`) is not something that the user even needs to bother with (or we can still have it in `values.yaml` under `conjurSslCertificateName`.

Reply:
Good call and I like your idea... Are we okay though with saving secrets (like ssl) in files?

Reply:
one clarification, `values.yaml` is for default parms we ship to customers. `custom-values.yaml` is customer supplied and used for overriding the defaults. Did you mean 
> How about letting the customer add the ssl_cert in... `custom-values.yaml`?

Reply:
how will the customer provide this secrets-provider-environment.conjurSslCertificate to custom-values.yaml ? If you will just put it there it is not secure. You can provide it via helm command line (there is a way to provide Values in command line),.

Reply:
yes, in that case then `custom-values.yaml` is better. Is this what we expect to have required values that do not have default values?

Reply:
> Are we okay though with saving secrets (like ssl) in files?

well, that's something to check with @shaharglazner

Reply:
> yes, in that case then `custom-values.yaml` is better. Is this what we expect to have required values that do not have default values?

it can be done for every values - not only for those which don't have defaults.

Reply:
> how will the customer provide this secrets-provider-environment.conjurSslCertificate to custom-values.yaml ? If you will just put it there it is not secure. You can provide it via helm command line (there is a way to provide Values in command line)

So you are saying not to have the customer copy + paste to custom-value, rather use a helm command for populating?

Reply:
mostly yes. the  cases that values.yaml parameters are passed via command line are usually the cases when we don't want the used to see / store them. for example:
helm install abc -set certificate=calculateCert() secret-provider-helm.tgz

Reply:
i think this is a good solution

Reply:
@oburstein-hub the ssl is not private.. it is the public SSL certificate value so it sitting in custom-values of the customer is not a security concern. @shaharglazner 
See here for explanation: https://docs.cyberark.com/Product-Doc/OnlineHelp/AAM-DAP/Latest/en/Content/Integrations/Kubernetes_deployApplicationsConjur-k8s-Secrets.htm?Highlight=secrets%20provider

Reply:
in this case it can be in custom-values.yaml

Reply:
@sigalsax @oburstein-hub afaiu the discussion, the certificate itself is a public piece of information (think of google.com certificate which can be retrieved without any permissions/restrictions). the sensitive information is the private key that signed that certificate, and should never leave the server.